x-pack/filebeat/input/azureblobstorage/input_test.go

```diff
@@ -395,24 +386,194 @@ func Test_StorageClient(t *testing.T) {
+				"max_workers":                         2000,
+				"poll":                                true,
+				"poll_interval":                       "10s",
+				"containers": []map[string]interface{}{
+					{
+						"name": mock.ConcurrencyContainer,
+					},
+				},
+			},
+			mockHandler: mock.AzureConcurrencyServer,
+			expectedLen: mock.TotalRandomDataSets,
+		},
+		{
+			name: "TestConcurrency_3000_Workers",
+			baseConfig: map[string]interface{}{
+				"account_name":                        "beatsblobnew",
+				"auth.shared_credentials.account_key": "7pfLm1betGiRyyABEM/RFrLYlafLZHbLtGhB52LkWVeBxE7la9mIvk6YYAbQKYE/f0GdhiaOZeV8+AStsAdr/Q==",
+				"max_workers":                         3000,
+				"poll":                                true,
+				"poll_interval":                       "10s",
+				"containers": []map[string]interface{}{
+					{
+						"name": mock.ConcurrencyContainer,
+					},
+				},
+			},
+			mockHandler: mock.AzureConcurrencyServer,
+			expectedLen: mock.TotalRandomDataSets,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			serv := httptest.NewServer(tt.mockHandler())
+			t.Cleanup(serv.Close)
+
+			cfg := conf.MustNewConfigFrom(tt.baseConfig)
+			conf := config{}
+			err := cfg.Unpack(&conf)
+			assert.NoError(t, err)
+			input := azurebsInput{
+				config:     conf,
+				serviceURL: serv.URL + "/",
+			}
+			name := input.Name()
+			if name != "azure-blob-storage" {
+				t.Errorf(`unexpected input name: got:%q want:"azure-blob-storage"`, name)
+			}
+
+			var src cursor.Source
+			// This test will always have only one container
+			for _, c := range input.config.Containers {
+				container := tryOverrideOrDefault(input.config, c)
+				src = &Source{
+					AccountName:   input.config.AccountName,
+					ContainerName: c.Name,
+					MaxWorkers:    *container.MaxWorkers,
+					Poll:          *container.Poll,
+					PollInterval:  *container.PollInterval,
+				}
+			}
+			v2Ctx, cancel := newV2Context()
+			t.Cleanup(cancel)
+			v2Ctx.ID += tt.name
+			var client publisher
+			st := newState()
+
+			var g errgroup.Group
+			g.Go(func() error {
+				return input.run(v2Ctx, src, st, &client)
+			})
+			timeout := time.NewTimer(100 * time.Second)
+			t.Cleanup(func() { timeout.Stop() })
+			client.Channel = make(chan beat.Event)
+			var receivedCount int
+		wait:
+			for {
+				select {
+				case <-timeout.C:
+					t.Errorf("timed out waiting for %d events", tt.expectedLen)
+					cancel()
+					return
+				case got := <-client.Channel:
+					_, err = got.Fields.GetValue("message")
+					assert.NoError(t, err)
+					receivedCount += 1
+					if receivedCount >= tt.expectedLen {
+						cancel()
+						break wait
+					}
+				}
+			}
+		})
+	}
+}
+
+type publisher struct {
+	Channel chan beat.Event
+	mu      sync.Mutex
+	cursors []map[string]interface{}
+}
+
+func (p *publisher) Publish(e beat.Event, cursor interface{}) error {
+	p.mu.Lock()
+	p.Channel <- e
```

Comment:
@efd6 I've added a custom client to the test but now, even with the concurrency fixes the there seems to be a data race detected here line 550. Any suggestions?

Reply:
The `Channel` field is being set after the errgroup is started. So you can fix this by moving the assignment up.
```diff
diff --git a/x-pack/filebeat/input/azureblobstorage/input_test.go b/x-pack/filebeat/input/azureblobstorage/input_test.go
index 8d7d49ac9f..f398ab4890 100644
--- a/x-pack/filebeat/input/azureblobstorage/input_test.go
+++ b/x-pack/filebeat/input/azureblobstorage/input_test.go
@@ -507,7 +507,9 @@ func Test_Concurrency(t *testing.T) {
                        v2Ctx, cancel := newV2Context()
                        t.Cleanup(cancel)
                        v2Ctx.ID += tt.name
-                       var client publisher
+                       client := publisher{
+                               Channel: make(chan beat.Event),
+                       }
                        st := newState()

                        var g errgroup.Group
@@ -516,7 +518,6 @@ func Test_Concurrency(t *testing.T) {
                        })
                        timeout := time.NewTimer(100 * time.Second)
                        t.Cleanup(func() { timeout.Stop() })
-                       client.Channel = make(chan beat.Event)
                        var receivedCount int
                wait:
                        for {
```
(note though that `Channel` doesn't need to be exported, I suggest `events`)

Reply:
I'm not a fan of the channel-based collection pattern that's used here. It's overly complex for what is needed. This is a sketch of what I would do (needs clean-up).
```diff
diff --git a/x-pack/filebeat/input/azureblobstorage/input_test.go b/x-pack/filebeat/input/azureblobstorage/input_test.go
index 8d7d49ac9f..288e17d929 100644
--- a/x-pack/filebeat/input/azureblobstorage/input_test.go
+++ b/x-pack/filebeat/input/azureblobstorage/input_test.go
@@ -507,7 +507,13 @@ func Test_Concurrency(t *testing.T) {
                        v2Ctx, cancel := newV2Context()
                        t.Cleanup(cancel)
                        v2Ctx.ID += tt.name
-                       var client publisher
+                       client := publisher{
+                               stop: func(e []beat.Event) {
+                                       if len(e) >= tt.expectedLen {
+                                               cancel()
+                                       }
+                               },
+                       }
                        st := newState()

                        var g errgroup.Group
@@ -516,38 +522,30 @@ func Test_Concurrency(t *testing.T) {
                        })
                        timeout := time.NewTimer(100 * time.Second)
                        t.Cleanup(func() { timeout.Stop() })
-                       client.Channel = make(chan beat.Event)
-                       var receivedCount int
-               wait:
-                       for {
-                               select {
-                               case <-timeout.C:
-                                       t.Errorf("timed out waiting for %d events", tt.expectedLen)
-                                       cancel()
-                                       return
-                               case got := <-client.Channel:
-                                       _, err = got.Fields.GetValue("message")
-                                       assert.NoError(t, err)
-                                       receivedCount += 1
-                                       if receivedCount >= tt.expectedLen {
-                                               cancel()
-                                               break wait
-                                       }
-                               }
+                       select {
+                       case <-timeout.C:
+                               t.Errorf("timed out waiting for %d events", tt.expectedLen)
+                               cancel()
+                       case <-v2Ctx.Cancelation.Done():
+                       }
+                       g.Wait()
+                       if len(client.events) < tt.expectedLen {
+                               t.Errorf("failed to get all events: got:%d want:%d", len(client.events), tt.expectedLen)
                        }
                })
        }
 }

 type publisher struct {
-       Channel chan beat.Event
+       stop    func([]beat.Event)
+       events  []beat.Event
        mu      sync.Mutex
        cursors []map[string]interface{}
 }

 func (p *publisher) Publish(e beat.Event, cursor interface{}) error {
        p.mu.Lock()
-       p.Channel <- e
+       p.events = append(p.events, e)
        if cursor != nil {
                var c map[string]interface{}
                chkpt, ok := cursor.(*Checkpoint)
@@ -565,6 +563,7 @@ func (p *publisher) Publish(e beat.Event, cursor interface{}) error {

                p.cursors = append(p.cursors, c)
        }
+       p.stop(p.events)
        p.mu.Unlock()
        return nil
 }
```
Note that the test still times out reasonably often when running with `-race` in the 3000 workers case.